nexus/src/external_api/client_api.rs

```diff
@@ -0,0 +1,257 @@
+use dropshot::{
+    endpoint, HttpError, HttpResponseOk, Query, RequestContext, TypedBody,
+};
+use http::{header, Response, StatusCode};
+use hyper::Body;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use serde_urlencoded;
+use std::sync::Arc;
+use uuid::Uuid;
+
+// Token granting à la RFC 8628 (OAuth 2.0 Device Authorization Grant)
+
+/// OAuth 2.0 error responses use 400 (Bad Request) with specific `error`
+/// parameter values to indicate protocol errors (see RFC 6749 §5.2).
+/// This is different from Dropshot's error `message` parameter, so we
+/// need a custom response builder.
+fn build_oauth_response<T>(
+    status: StatusCode,
+    body: &T,
+) -> Result<Response<Body>, HttpError>
+where
+    T: ?Sized + Serialize,
+{
+    let body = serde_json::to_string(body)
+        .map_err(|e| HttpError::for_internal_error(e.to_string()))?;
+    Ok(Response::builder()
+        .status(status)
+        .header(header::CONTENT_TYPE, "application/json")
+        .body(body.into())?)
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+pub struct AuthenticateParams {
+    pub client_id: Uuid,
+}
+
+/// Start an OAuth 2.0 Device Authorization Grant
+///
+/// This endpoint is designed to be accessed from an unauthenticated
+/// API client. It generates and records a `device_code` and `user_code`
+/// which must be verified and confirmed prior to a token being granted.
+#[endpoint {
+    method = POST,
+    path = "/client/authenticate",
+    content_type = "application/x-www-form-urlencoded",
+    tags = ["hidden"], // "token"
+}]
+pub async fn client_authenticate(
+    rqctx: Arc<RequestContext<Arc<ServerContext>>>,
+    params: TypedBody<AuthenticateParams>,
+) -> Result<Response<Body>, HttpError> {
+    let apictx = rqctx.context();
+    let nexus = &apictx.nexus;
+    let params = params.into_inner();
+    let handler = async {
+        let opctx = nexus.opctx_external_authn();
+        let request = rqctx.request.lock().await;
+        let host = match request.headers().get(header::HOST) {
+            None => {
+                return build_oauth_response(
+                    StatusCode::BAD_REQUEST,
+                    &serde_json::json!({
+                        "error": "invalid_request",
+                        "error_description": "missing Host header",
+                    }),
+                )
+            }
+            Some(host) => match host.to_str() {
+                Ok(host) => host,
+                Err(e) => {
+                    return build_oauth_response(
+                        StatusCode::BAD_REQUEST,
+                        &serde_json::json!({
+                            "error": "invalid_request",
+                            "error_description": format!("could not decode Host header: {}", e)
+                        }),
+                    )
+                }
+            },
+        };
+
+        let model = nexus.client_authenticate(&opctx, params.client_id).await?;
```

Comment:
Is this going to do the right thing in terms of the error response form?  Say the database is down.  It seems like this will generate an `Err(Error::ServiceUnavailable)`, which will get turned into a `dropshot::HttpError`, which I think we're _not_ supposed to be returning from this function?

If that's true, even if we fix it here, it seems easy to accidentally re-introduce that bug.  I wonder if we should have the body of these handlers call a Rust function that returns only `Response<Body>` (no `Result`).  We could document why -- we need to format all the responses ourselves to ensure compliance.

Reply:
Excellent catch, thank you. I'd like to take this as a follow-up, recorded as #1286.